libfuse/dir.go

```diff
@@ -681,35 +682,35 @@ func (d *Dir) Rename(ctx context.Context, req *fuse.RenameRequest,
 		if err != nil {
 			return err
 		}
+	case *FolderList, *Root:
+		// This normally wouldn't happen since a presumably pre-check on
+		// destination permissions would have failed. But in case it happens, it
+		// should be a EACCES according to rename() man page.
+		return fuse.Errno(syscall.EACCES)
 	default:
-		// The destination is not a TLF instance, probably
-		// because it's Root (or some other node type added
-		// later). The kernel won't let a rename newDir point
-		// to a non-directory.
-		//
-		// We have no cheap atomic rename across folders, so
-		// we can't serve this. EXDEV makes `mv` do a
-		// copy+delete, and the Lookup on the destination path
-		// will decide whether it's legal.
-		return fuse.Errno(syscall.EXDEV)
-	}
-
-	if d.folder != realNewDir.folder {
-		// Check this explicitly, not just trusting KBFSOps.Rename to
-		// return an error, because we rely on it for locking
-		// correctness.
-		return fuse.Errno(syscall.EXDEV)
-	}
-
-	// overwritten node, if any, will be removed from Folder.nodes, if
-	// it is there in the first place, by its Forget
-
-	if err := d.folder.fs.config.KBFSOps().Rename(
-		ctx, d.node, req.OldName, realNewDir.node, req.NewName); err != nil {
-		return err
+		// This shouldn't happen unless we add other nodes. EIO is not in the error
+		// codes listed in rename(), but there doesn't seem to be any suitable
+		// error code listed for this situation either.
+		return fuse.Errno(syscall.EIO)
 	}
 
-	return nil
+	err = d.folder.fs.config.KBFSOps().Rename(ctx,
+		d.node, req.OldName, realNewDir.node, req.NewName)
+
+	switch err.(type) {
+	case nil:
+		return nil
+	case libkbfs.RenameAcrossDirsError:
+		epath, er := sysutils.GetExecPathFromPID(req.Pid)
+		if er != nil {
+			d.folder.fs.log.CDebugf(ctx,
+				"Dir Rename: getting exec path for PID %d error: %v", req.Pid, er)
+			return err
+		}
+		return libkbfs.ExdevForUnsupportedApplicationError{ExecPath: epath}
```

Comment:
The comment for `ExdevForUnsupportedApplicationError` says that the error should only be used for programs that are known not to respect EXDEV.  But this code doesn't check anything about the program.  Is the idea that the desktop app will check?  If so, I guess that comment should be updated.  If not, we need a list of unsupported applications to check against here.